### ISSUE TITLE
Bug: Connection terminated unexpectedly

### ISSUE THREAD

**Issue by rafamel** [knex/knex#3523]
# Environment

Knex version: 0.20.1
Database + version: postgres:11.5
OS: Alpine Linux

Additional information:
- Database: I've confirmed the bug to happen in *postgres* versions at least up from `9.5`. The database is running on a container, currently [`postgres:11.5`](https://hub.docker.com/_/postgres).
- Knex environment: Knex is currently running on node v12, though I've confirmed the bug to happen on v8 and v10 too. It's running on a container, currently [`node:12.13.0-alpine`](https://hub.docker.com/_/node). The current `pg` version is 7.12.1; though I've seen this happen with previous versions, I haven't documented this bit.
- Knex configuration object:
```
({
  client: 'pg',
  connection: 'postgres://user:password@server:5432/db',
  pool: { min: 2, max: 10 }
})
```

# Bug

I am getting a `Connection terminated unexpectedly` error in production, after long periods of inactivity (think +1h) during which no queries are made, causing the first query after that period to fail. Any query made after that first failing query will succeed.

This was first reported on issue #3258. In that issue, it was reported happening on knex 0.16.5, so this has been going on at least since then. It was also stated that a downgrade to 0.13.0 fixes the issue, which met the suggestion that maybe newer knex versions logged the error while previous ones didn't. I've confirmed this **not** to be the case. Downgrading to knex 0.13.0 does in fact fix the issue, with queries being successful after long periods of inactivity.

A typical error message will look like this:
- message: `select \"user\".* from \"user\" where \"email\" ilike $1 - Connection terminated unexpectedly`
- stack: `Error: Connection terminated unexpectedly\n at Connection.<anonymous> (/usr/src/app/node_modules/pg/lib/client.js:252:9)\n at Object.onceWrapper (events.js:299:28)\n at Connection.emit (events.js:210:5)\n at Socket.<anonymous> (/usr/src/app/node_modules/pg/lib/connection.js:77:10)\n at Socket.emit (events.js:210:5)\n at TCP.<anonymous> (net.js:658:12)`

[`node-postgres` #1324](https://github.com/brianc/node-postgres/issues/1324) might help providing additional insight on the matter.

**Comment by elhigu** [comment]
Would be good if you can provide test code to verify that. For example simple example project with docker-compose file and knex app that waits long enough to encounter that error.

We have got various reports of that bug, but no-one has been able to show if the problem is in user's code or in knex. 

If you can catch that error with env variable DEBUG=knex:* set it would also be helpful to verify that the error is not thrown from transaction that has been left open accidentally etc.

**Comment by rafamel** [comment]
I have good news! After further investigating the issue, I believe I can point towards a solution -I'll leave that for last, as there's a bit to unpack.

Unfortunately, I haven't been able to reproduce the error locally, even with the same setup and compose configuration as the remote server. It is quite a strange thing, as the only difference I can appreciate between both environments is I'm running `docker-compose` locally, while the server is running a docker swarm. To reiterate, I've set up a repo that reproduces this error on a remote production server, but (somehow) not locally. I will post the logs below (last), so for the purpose of understanding them, I've made the repo available [here](https://github.com/rafamel/knex-error). As you can see there, I'm making a first successful query right after the process starts, making a 2nd query after 20 minutes (which fails), and a 3rd query immediately after (which succeeds).

Some additional info, just for the sake of future reference:

- It seems the idle time needed for the connection error to happen after querying is somewhere between 15 and 20 minutes. Queries after 15 minutes of inactivity will succeed; while after 20, they will fail.
- knex 0.13.0 doesn't show this issue, while every version after, including 0.14.0, does. However, on 0.14.0 a new connection won't be acquired on the next query, so they will fail indefinitely, which is solved on 0.15.0.

And now, for the useful bit: it turns out `pg` is not emitting the connection error until *after* the query has been made, but before it has resolved. This is the sequence of events:

1. *knex* public api query is made.
2. *knex*'s `dialects/postgres/index.js` `_query` is called with a *pg* `connection`, which calls *pg*'s `connection.query`.
3. *pg*'s `connection` emits an `'error'` event, which is received by `connection.on('error', cb)` at *knex*'s `dialects/postgres/index.js`.
4. *pg*'s `connection.query` calls its callback with the same error that had just been emitted, hence *knex*'s `dialects/postgres/index.js` `_query` rejects with said error.
6. *pg*'s `connection` emits an `'end'` event, which is received by `connection.on('end', cb)` at *knex*'s `dialects/postgres/index.js`.
7. *knex* releases connection to pool.
8. The next time *knex* receives a query to be made, it will acquire a new *pg* connection (`acquireRawConnection`) and succeed.

So essentially, it could be argued that *maybe* there's an issue with the containerized production environment/setup, and/or with *pg,* which is not reporting a connection termination until the time at which a query is made -though I don't know whether reporting it earlier would be feasible. But regardless, there's a (very much fixable) issue with *knex:* if a connection errors out in the time that goes between a query is made and its resolution, the query will fail. I would imagine this would be the case not just with postgres, but with other dialects too.

Hence I would suggest that, in this event, a new connection is acquired and the query retried. It might make sense to pass that error (make the query fail with it), though, in the event of the connection having being acquired just before it was made. So, to recap, I'd propose that if a query errors out and the connection was just acquired for the purpose of the query, it rejects with the error, but otherwise it checks whether the connection has failed, and if so, acquires a new connection and retries the query.

Last, the logs. These are for v0.20.1, but I have logs for every minor version starting on 0.13.0.

- First query, right after the process starts:

```
2019-11-09T09:22:02.057558944Z Elapsed: 0 ms
2019-11-09T09:22:02.058104661Z 2019-11-09T09:22:02.057Z knex:client acquired connection from pool: __knexUid1
2019-11-09T09:22:02.059072595Z {
2019-11-09T09:22:02.059140593Z method: 'select',
2019-11-09T09:22:02.059150803Z options: {},
2019-11-09T09:22:02.059156884Z timeout: false,
2019-11-09T09:22:02.059162835Z cancelOnTimeout: false,
2019-11-09T09:22:02.059169678Z bindings: [ 'example@example.com' ],
2019-11-09T09:22:02.059176932Z __knexQueryUid: '0ab6fbe6-f4d0-497a-87b0-057033964950',
2019-11-09T09:22:02.059183233Z sql: 'select * from "user" where "email" = ?'
2019-11-09T09:22:02.059191860Z }
2019-11-09T09:22:02.059483679Z 2019-11-09T09:22:02.059Z knex:query select * from "user" where "email" = ? trx2
2019-11-09T09:22:02.059543883Z 2019-11-09T09:22:02.059Z knex:bindings [ 'example@example.com' ] trx2
2019-11-09T09:22:02.063050947Z 2019-11-09T09:22:02.062Z knex:client releasing connection to pool: __knexUid1
2019-11-09T09:22:02.063380849Z Result []
```

- Second query, after 20 minutes:
```
2019-11-09T09:42:02.067012525Z Elapsed: 1200000 ms
2019-11-09T09:42:02.071066942Z 2019-11-09T09:42:02.069Z knex:client acquired connection from pool: __knexUid1
2019-11-09T09:42:02.071878431Z {
2019-11-09T09:42:02.071909690Z method: 'select',
2019-11-09T09:42:02.071916663Z options: {},
2019-11-09T09:42:02.071921983Z timeout: false,
2019-11-09T09:42:02.071926913Z cancelOnTimeout: false,
2019-11-09T09:42:02.071931722Z bindings: [ 'example@example.com' ],
2019-11-09T09:42:02.071936431Z __knexQueryUid: '229e5c4a-9ed3-4d82-a25e-b048e7cb9c0d',
2019-11-09T09:42:02.071941269Z sql: 'select * from "user" where "email" = ?'
2019-11-09T09:42:02.071946579Z }
2019-11-09T09:42:02.072055515Z 2019-11-09T09:42:02.071Z knex:query select * from "user" where "email" = ? trx2
2019-11-09T09:42:02.072125507Z 2019-11-09T09:42:02.072Z knex:bindings [ 'example@example.com' ] trx2
2019-11-09T09:42:02.081042312Z 2019-11-09T09:42:02.080Z knex:client releasing connection to pool: __knexUid1
2019-11-09T09:42:02.082578957Z Error: Connection terminated unexpectedly
2019-11-09T09:42:02.082619874Z at Connection.<anonymous> (/usr/src/app/node_modules/pg/lib/client.js:252:9)
2019-11-09T09:42:02.082627337Z at Object.onceWrapper (events.js:299:28)
2019-11-09T09:42:02.082633188Z at Connection.emit (events.js:210:5)
2019-11-09T09:42:02.082638899Z at Socket.<anonymous> (/usr/src/app/node_modules/pg/lib/connection.js:77:10)
2019-11-09T09:42:02.082645932Z at Socket.emit (events.js:210:5)
2019-11-09T09:42:02.082650612Z at TCP.<anonymous> (net.js:658:12)
```

- Third query, right after the 2nd:

```
2019-11-09T09:42:02.084190783Z Elapsed: 0 ms
2019-11-09T09:42:02.084868650Z Connection Error: Connection ended unexpectedly
2019-11-09T09:42:02.097410334Z 2019-11-09T09:42:02.097Z knex:client acquired connection from pool: __knexUid3
2019-11-09T09:42:02.098051381Z {
2019-11-09T09:42:02.098071179Z method: 'select',
2019-11-09T09:42:02.098076118Z options: {},
2019-11-09T09:42:02.098080576Z timeout: false,
2019-11-09T09:42:02.098085165Z cancelOnTimeout: false,
2019-11-09T09:42:02.098089503Z bindings: [ 'example@example.com' ],
2019-11-09T09:42:02.098094572Z __knexQueryUid: '6e702fd9-b0d5-4c8e-9ee7-0b6404fec6d6',
2019-11-09T09:42:02.098099522Z sql: 'select * from "user" where "email" = ?'
2019-11-09T09:42:02.098104000Z }
2019-11-09T09:42:02.098367957Z 2019-11-09T09:42:02.098Z knex:query select * from "user" where "email" = ? undefined
2019-11-09T09:42:02.098583893Z 2019-11-09T09:42:02.098Z knex:bindings [ 'example@example.com' ] undefined
2019-11-09T09:42:02.103675485Z 2019-11-09T09:42:02.103Z knex:client releasing connection to pool: __knexUid3
2019-11-09T09:42:02.104059078Z Result []
```

**Comment by elhigu** [comment]
Thank you very much. Those findings are really useful. Looks like things are happening as I expected so that error is not either happening or registered before it is taken away from pool and data is actually sent over the connection.

One can probably work around this issue by setting minimum pool size to 0 and set idle timeout to be like 5 minutes or so. In that way connections will be discarded from pool before that.

> But regardless, there's a (very much fixable) issue with knex: if a connection errors out in the time that goes between a query is made and its resolution, the query will fail. 

Actually connection error might happen at any time (or any other surprise error) and app should be written to handle unexpected errors causing queries to fail. Automatic retrying will not resolve it, except for some very specific cases, where connection error gets fixed automatically in some time frame and when transactions are not used.

So knex is intentionally not trying to retry failed queries automatically. Retrying would also cause other problems, like a delay for app to be able to notice that your DB connection is down (and after noticing you would still have to have that error management for unexpectedly failed query).

**Comment by rafamel** [comment]
Hi, thank you for taking the time ;) In general, I can see where you're coming from, however:

> Actually connection error might happen at any time (or any other surprise error) and app should be written to handle unexpected errors causing queries to fail. Automatic retrying will not resolve it, except for some very specific cases, where connection error gets fixed automatically in some time frame and when transactions are not used.

I'm failing to get the point here. When a connection error occurs, knex will discard it and acquire a new one, hence there's a expectation that connection errors (not other kind of query errors, but yes, connection errors) will be handled, when possible to reconnect, by knex. This case fits there, with the only difference that the connection happens to emit the error after the query has been made (but before it's been resolved). Hence my proposal to log the connection error, discard the query error, get a new connection, and retry the query -not for every case, but only for this particular case in which the connection has emitted an error after querying but before resolving. This seems the most consistent to me with the current behavior otherwise, given the above expectation.

> Retrying would also cause other problems, like a delay for app to be able to notice that your DB connection is down (and after noticing you would still have to have that error management for unexpectedly failed query).

On most cases, I'd assume that the dialect's `connection` will emit a connection error at any time (so not during an ongoing query), in which case, the current behavior applies, so no harm done. As pointed out, this would only apply in cases when the connection error is emitted during a query, hence both because the # of times that would happen by random chance is statistically very low, and because the delay would be realistically unnoticeable (given it's not something that accumulates, but rather a singularly occurring event), I'd say the downside would be more than justified.

> One can probably work around this issue by setting minimum pool size to 0 and set idle timeout to be like 5 minutes or so. In that way connections will be discarded from pool before that.

It might be sensible to set defaults that avoid this issue or, at least, document it in some way. But regardless, I am somewhat confused with the current defaults. I'm assuming when you say `set idle timeout to be like 5 minutes or so` you're referring to `tarn.js` `idleTimeoutMillis`, to be set in the `pool` key object within knex's configuration object. Unless `knex` is passing a different default value (is it?), [`idleTimeoutMillis` seems to default to 30000ms](https://github.com/Vincit/tarn.js/blob/master/src/Pool.ts), 30 seconds, which should have avoided this error altogether. Any insight on this?

**Comment by elhigu** [comment]
I have explained why knex does not retry on connection errors or on any other errors so many times to so many people that I should probably just find a link to some older discussion 😬

> 'm failing to get the point here. When a connection error occurs, knex will discard it and acquire a new one, hence there's a expectation that connection errors (not other kind of query errors, but yes, connection errors) will be handled, when possible to reconnect, by knex.

If there is a connection error when connection is in pool we know that it is not used by anyone so it is owned by knex and we can do what ever we want to with it (actually we don't discard it until someone tries to get it from pool and validation phase sees that connection was broken while being in the pool).

Actually now that I think of it if I recall correctly even pool does not retry for connection errors. So if connection is broken when connection was in pool, then it will be discarded and fresh connection is returned instead, but if that fresh connection fails because of connection error, it will not be retried, but the error is passed to the user. This is also for the same reason, because we don't want knex to hide those connection errors which happen during queries.

> Hence my proposal to log the connection error, discard the query error, get a new connection, and retry the query -not for every case, but only for this particular case in which the connection has emitted an error after querying but before resolving. This seems the most consistent to me with the current behavior otherwise, given the above expectation.

It doesn't make sense to make code more complicated and add single retry for this single case. Expectation is that when connection is got from pool it is usable... if there is connection error happening afterwards, then it is passed to the user. We don't want to mask errors, you still can have them coming at any time which is not even related to this issue and then you would need to retry for that case too (because it will be hard to distinct from other case). Also as I mentioned above you would still need to have error handling for the case where connection error comes and it takes longer that just one query or if it happens during transaction.

For this particular case I would love to find fix the root cause of the problem, but until then setting pool min to 0 should do just fine. It is pretty uncommon case where database is without use for hours and that you would still want to have pool populated with connections during that idle time.

> On most cases, I'd assume that the dialect's connection will emit a connection error at any time (so not during an ongoing query), in which case, the current behavior applies, so no harm done. As pointed out, this would only apply in cases when the connection error is emitted during a query, hence both because the # of times that would happen by random chance is statistically very low, and because the delay would be realistically unnoticeable (given it's not something that accumulates, but rather a singularly occurring event), I'd say the downside would be more than justified.

I really didn't understand what you tried to say with that. Only benefit that I see with having that retrying logic is that it will work as a hack to kind of fix something that does not seem to be a knex bug. Preventing pool from keeping idle connections there for long periods of time is a lot better workaround for this (happens with some very specific environment).

> Unless knex is passing a different default value (is it?), idleTimeoutMillis seems to default to 30000ms, 30 seconds, which should have avoided this error altogether. Any insight on this?

Pool min size 0 was the key part of that work around. If pool size min is bigger that 0 then idle timeout will not do anything after pool has already reached its minimum size.

**Comment by rafamel** [comment]
Hi, thank you for taking the time of explaining the reasoning for this (yet again). I know it can get tiring; thank you.

From your previous statements I got the idea that passing errors and not retrying queries was intentional, but not necessarily not handling connection errors. Your latest explanation makes it clear that it is your intent that connection errors are explicitly managed by the user in all cases ("if connection is broken when connection was in pool, then it will be discarded and fresh connection is returned instead, but if that fresh connection fails because of connection error, it will not be retried, but the error is passed to the user"). I had previously assumed this wasn't the case. With that in mind, I think what you're saying is consistent with the expectations set, hence no changes should be made, as regardless of whether they happen during or before a query, they will be passed either way.

However, it might be that I didn't look carefully enough, but I think it would be very good if this information regarding the handling of broken connections vs. errored connections **had a prominent place in knex's documentation** in order for devs to have certainty on the expectations. On a side note, I also couldn't locate the default values for the configuration options in the docs, which would also be helpful.

Anyhow, I have confirmed that setting the `pool.min` to `0` prevents this from happening. I wonder though, why is it `0` not the default value?

**Comment by ctusch** [comment]
I'm having the same problem. It's very hard to reproduce since it's only occurring on the system of my remote colleagues (no docker). I can run the server here and have it idle for hours and it still doesn't happen. Also no signs of it on our dockerized ECS instance. They all connect to the same Postgres server. I'll try the workaround.

**Comment by elhigu** [comment]
> However, it might be that I didn't look carefully enough, but I think it would be very good if this information regarding the handling of broken connections vs. errored connections had a prominent place in knex's documentation in order for devs to have certainty on the expectations. On a side note, I also couldn't locate the default values for the configuration options in the docs, which would also be helpful. 

Certainly that would be good to have 👍

**Comment by jamesdixon** [comment]
@rafamel I'm running Knex 0.20.7 and I'm still seeing this occur occasionally even with `pool.min = 0`. Was there anything else you did to work around this? 

I read in another thread that handling the error is good enough just so it doesn't crash the app, which is why I'm assuming this wasn't an issue in older versions of Knex.

**Comment by elhigu** [comment]
@jamesdixon maybe there really are some connection errors occuring. You need to be able to pinpoint if those messages are coming from connections that you are using in application code or from those which are actually still reserved in pool. However those errors should not crash the app. 

You could open a new issue with more information, when and what is exactly happenin in your case and then someone might be able to help in your particular case.

**Comment by jamesdixon** [comment]
@elhigu thanks. These issues are occurring in a Lambda environment so it's certainly a possibility there's an underlying issue. I haven't had the best of luck with Knex and pooling in AWS.

**Comment by iw2rmb** [comment]
Same here. Local instance of nodejs12 + pg 12 on docker. Connection terminated if not used for approx 1 hour.

**Comment by elhigu** [comment]
@iw2rmb please send test code for that... for example docker-compose file would be helpful. This problem has not been able to be reproduced outside of AWS lambda so this would make a huge difference for being able to resolve the bug. Though I'm pretty sure that error you are experiencing is actually not the same issue.... Is that terminated connection returned from pool or is it automatically refreshed before next query after 12 hours is sent to DB?

**Comment by NelsonFrancisco** [comment]
I'm experiencing similar issue with in a docker swarm cluster.
`select "users".* from "users" where "email" = $1 limit $2 - Connection terminated unexpectedly`
Using postgres 12, node 12.3.1 and knex 0.19.5.
knex configuration pool connections min:2, max:10

I have Sentry set up in my environment. I can match the timing of the error with several of my docker services.
postgres service reports nothing.
my api service logs knex connections.
Here's the log at that same time:
```
2020-03-12T13:04:18.716Z knex:client acquired connection from pool: __knexUid1
2020-03-12T13:04:18.716Z knex:query select "users".* from "users" where "email" = ? limit ? trx
12
2020-03-12T13:04:18.716Z knex:bindings [ 'admin@example.com', 2 ] trx12
2020-03-12T13:04:18.719Z knex:client releasing connection to pool: __knexUid1
2020-03-12T13:04:37.964Z pg-listen:paranoid Paranoid connection check ok
```
It looks like the query was correctly executed, right?

**Comment by elhigu** [comment]
@NelsonFrancisco I don't see anythin in logs which would refer that query was ran successfully... It could have as well got broken connection from pool try to run query and then return it.

**Comment by NelsonFrancisco** [comment]
I was assuming that if an error occurred, knex would report it.

**Comment by elhigu** [comment]
If query fails for some reason, it is just normal behaviour and not specially logged. If you like to log results / errors you need to setup for example knex.on('query-error',... ) and knex.on('query-response',... ) event handlers.

**Comment by elhigu** [comment]
you application logic should anyways have proper error handling for those cases, for example with REST API return some 500 error.

**Comment by NelsonFrancisco** [comment]
I agree. Currently, it is timing out, and returning 502 bad request. But it is very strange.
I'm using objection with knex. Every error that happens when I execute a query is properly catch in a middleware - and ultimately, if the error is not known, a fallback middleware returns 500.

I'm pretty sure the exception is not being thrown when I "await" for objection's "where" method.i.e, when I await knex's "select" method. Because if it would, then the API would catch the error. So, maybe the promise is taking too long OR it is being resolved and the error happens asynchronously

**Comment by huchenhai** [comment]
I'm getting the same problem in docker and Kubernetes environment. 
When trying to connect to Postgres,  the first call failed, and then request immediately next call works.
{ Error: read ECONNRESET
    at TLSWrap.onStreamRead (internal/stream_base_commons.js:111:27) errno: 'ECONNRESET', code: 'ECONNRESET', syscall: 'read' }
Connection Error: Connection ended unexpectedly

**Comment by NelsonFrancisco** [comment]
> I'm getting the same problem in docker and Kubernetes environment.
> **The first call failed if few hours not use, and the next request works.**
> Connection Error: Connection ended unexpectedly

Similar issue happens to me. First call in a long time fails. I can't exactly say the time interval.

> app.get('/all', function (req, res, next) {
> knex.select().table('main').then(data => {
> res.json(data);
> }).catch(next);
> });

In your case, is "next" callback executed?

**Comment by huchenhai** [comment]
@NelsonFrancisco  yes, the second call executed successfully.
The connection timeout interval here seems to be 20 min

**Comment by huchenhai** [comment]
@jamesdixon  guess James is the most experienced person on this issue

**Comment by NelsonFrancisco** [comment]
That's good, because the same happens to me

**Comment by huchenhai** [comment]
@NelsonFrancisco  I've set the min pool size to 0,  did a test again this morning, it seems works. 
I don't see that error message from log again

**Comment by NelsonFrancisco** [comment]
> @NelsonFrancisco I've set the min pool size to 0, did a test again this morning, it seems works.
> I don't see that error message from log again

I can say, without a doubt, that fixed the problem.
One week and going without this issue

**Comment by elhigu** [comment]
I hid bunch of comments about a separate issue, when that connection closed error is shown, but which didn't  cause any queries to fail (cases where broken connections were actually never returned from pool).

**Comment by david-wb** [comment]
Any updated on this? I see this issue has been open for a year. 

I'm facing the same problem while running a nodejs app in google cloud-run. 

Queries are failing in production when the service has been left idle for a long time. The error message is 
```
select ... - Connection terminated unexpectedly
```

Here are my connection pool settings for Knex.
```
pool: {
    min: 1,
    max: 5,
    acquireTimeoutMillis: 60000,
    idleTimeoutMillis: 600000,
}
```

**Comment by lynxtaa** [comment]
I've solved it by setting minimal pool size to zero: `pool: { min: 0, max: 10 }`

**Comment by david-wb** [comment]
@lynxtaa I changed my pool settings to the following and the error has not occurred in a couple days, so I guess this fixes the problem. I'll comment again if it resurfaces. Thanks.
```
pool: {
    min: 0,
    max: 5,
    acquireTimeoutMillis: 60000,
    idleTimeoutMillis: 600000,
}
```

**Comment by kamikazechaser** [comment]
Still an issue. I suggest that the docs be changed to reflect the suggestions given above.

My stack trace:

```json
{"level":50,"time":1614497262294,"pid":1,"hostname":"843348d00255","reqId":194,"errno":-110,"code":"ETIMEDOUT","syscall":"read","stack":"Error: read ETIMEDOUT\n    at TCP.onStreamRead (internal/stream_base_commons.js:209:20)","type":"Error","msg":"select...REDACTED_SQL_QUERY) - read ETIMEDOUT"}
```

**Comment by rizkyramadhan** [comment]
Yes this still happen to me... instead of setting pool, I think it's better if we have know how to catch time out and recover properly.

**Comment by ariden84** [comment]
you can reproduce the bug in any environment by entering

`    "pool": {
      "min": 0,
      "max": 1
    },
    "acquireConnectionTimeout": "10"`

a max pool to 1 or an acquireConnectionTimeout to 10ms and call several requests in a row

**Comment by NelsonFrancisco** [comment]
I got some timers running and making queries. First query fails if settings are as @ariden84 said
```
KnexTimeoutError: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?
    at Client_PG.acquireConnection (/home/whatever/node_modules/knex/lib/client.js:349:26)
    at runNextTicks (internal/process/task_queues.js:62:5)
    at listOnTimeout (internal/timers.js:518:9)
    at processTimers (internal/timers.js:492:7) {
  sql: 'select 1',
  bindings: undefined
}
```

**Comment by Zyles** [comment]
I got this problem today.

If I use pooling and let the app sit for a couple of hours, then make a request I get "Connection Error: Connection ended unexpectedly". After that first failing request the second request works.

**Comment by RodolfoSilva** [comment]
@Zyles, I've the same problem.

**Comment by Warkanlock** [comment]
I'm on the same boat as everyone here, first request fails but after that everything works as expected

**Comment by rizkyramadhan** [comment]
> I'm on the same boat as everyone here, first request fails but after that request, everything works as expected
> 
> Seems like `min : 0`gives us a solution given that `tarn.js` will remove every idle instance given that the minimum number of instances in the pool is determined by the `minKeep` and `maxDestroy` instantiate inside the `check()` method on the Pool object. (https://github.com/Vincit/tarn.js/blob/master/src/Pool.ts#L247)
> 
> So, passing a 0 **will not retain instances inside Pool object** and you won't have idle instances. So, this is more like a bandaid to the problem than actually a solution for it but works at the end since the `check()` function inside Pool.js on tarn.js will use this minimum size to keep instances alive and given that the minimum size is 0 will retain 0 instances alive

Do you have any idea how to fix this from the root cause ?

**Comment by Warkanlock** [comment]
> > I'm on the same boat as everyone here, first request fails but after that request, everything works as expected
> > Seems like `min : 0`gives us a solution given that `tarn.js` will remove every idle instance given that the minimum number of instances in the pool is determined by the `minKeep` and `maxDestroy` instantiate inside the `check()` method on the Pool object. (https://github.com/Vincit/tarn.js/blob/master/src/Pool.ts#L247)
> > So, passing a 0 **will not retain instances inside Pool object** and you won't have idle instances. So, this is more like a bandaid to the problem than actually a solution for it but works at the end since the `check()` function inside Pool.js on tarn.js will use this minimum size to keep instances alive and given that the minimum size is 0 will retain 0 instances alive
> 
> Do you have any idea how to fix this from the root cause ?. We can patch-package knex.js if we have a fix...

I was really packed at work that day, but I can try to dig into today and see what I can do

**Comment by Warkanlock** [comment]
I might have found something, correct me please If I'm wrong because this is my first time digging the code:

On the method ``acquireRawConnection()`` on ``postgres/index.js`` dialect, we are using an event handler for the event type "**end**" which I think is incorrect because we dispose of the error to the pipe think this cause the "idle" error we currently see (correct me here If I'm wrong and I'm not taking in count other use cases).

**``dialect/postgres/index.js line 78``**
```        
connection.on('end', (err) => {
          connection.__knex__disposed = err || 'Connection ended unexpectedly';
});
```

``pg.Client documentation`` (https://node-postgres.com/api/client)
```
client.on('end') => void
When the client disconnects from the PostgreSQL server it will emit an end event once.
```

Given docs from ``pg, `` this event is emitted **only on a client disconnection**, so what could be happening here is that we left the connection "open" on the pool but closed on the Postgres pool, so when we try to communicate with the connection using the ``pg`` driver we receive that error.

Aside from this, also we should extend the **validate connection** method inside the pool to check if the connection already exists or is it valid. As far as I saw on the code while digging, _correct me if this is on purpose_, the validate method for the pool config inside ``client.js`` is quite silly (not using connection object to check if it's valid)

![image](https://user-images.githubusercontent.com/13340320/122677628-7d85ae80-d1b9-11eb-867d-0ac406f655a3.png)

Maybe in conjunction with the ``end`` event listening commented on the first part of this comment, and the validate extension to something like the following could do the job:

```
  validateConnection(connection) {
    return connection && connection.connected;
  }
```

Currently, I'm trying to replicate it using the repository posted by @rafamel but without luck (I'm waiting for idle connections up to 20 minutes but nothing happens)

**Comment by rizkyramadhan** [comment]
Wow, this is great start. This week I don't have any bandwidth to do the digging, pardon. This bug is hard to replicate because we have to wait for the timeout, we don't have any reliable way to confirm whether this is fixed or not... Maybe we can take look another postgres library how they did the pool ?

**Comment by intech** [comment]
I know of an easy way to reproduce this error, since I forward the remote port via ssh from the server, then I just interrupt the forwarding and start reconnect. I suppose the same behavior will happen with a simple restart of the database.

**Comment by Warkanlock** [comment]
I could reproduce the error emulating a postgresql database based on @intech comments (but here it's 100% offline)

Steps to reproduce the bug:

1- Emulate and expose a Postgres instance using ``docker-compose``

```
version: '3.4'
services:
  db:
    image: 'postgres:11-alpine'
    ports: 
      - "5432:5432"
    environment:
      POSTGRES_USER: test
      POSTGRES_PASSWORD: test
      POSTGRES_DB: test
```

2- Run a Node.JS script with a ``setInternal()`` to hit the database constantly while **we are doing step 3**

```javascript

const knex = Knex({
  client: 'pg',
  connection: 'postgresql://dev:dev@localhost:5432/dev', //use the public URL not the docker one, in order to have more flexibility at the time of debugging 
  pool: { min: 1, max: 5 },
  debug: true
});


setInterval(async () => {
  await knex.select().from('some-table'); //you could use postgres common tables, not need to create a migration
}, 2000);
```

2- Run, as a superuser, this query (will shutdown all connections associated with the database **test**
```sql
SELECT
  pg_terminate_backend(pid)
FROM
  pg_stat_activity
WHERE
  pid <> pg_backend_pid()
  AND datname = 'test';
```

As you can see in the following picture, **after running the script of step 2, while in parallel, we are running the query of step 3,** we will get idle connections (dead ones) with the expected error message (Connection terminated unexpectedly). This is helpful to replicate and fix this issue

![image](https://user-images.githubusercontent.com/13340320/122692141-d5e49c80-d209-11eb-9f20-8fd19e420a7b.png)

**Comment by rizkyramadhan** [comment]
@Warkanlock use [patch-package](https://www.npmjs.com/package/patch-package).

**Comment by rizkyramadhan** [comment]
Wow, if you fixed it, can you share the .patch file in patches directory ? That file is generated by patch-package.

**Comment by kibertoad** [comment]
@Warkanlock Wouldn't it be better to fix the problem upstream in the library itself? What changes are you doing via that patch?

**Comment by Warkanlock** [comment]
> Wow, if you fixed it, can you share the .patch file in patches directory ? That file is generated by patch-package.

I didn't use ``patch-package`` but ``npm link`` instead, I'm doing a PR with the changes but before pushing something. I'm looking for advice here, quite unsure about the results. I don't know at 100% if this is the way to replicate the issue (despite the fact the errors return the same message and status) nor if this bug is the bug we all saw before. 

But here a demonstration of the connection reconnecting with the pool after an unexpected shutdown (or idle behaviour). As we can see, the first request hit correctly, the second one does not but at the end resolve either way since the pool is re-created

![image](https://user-images.githubusercontent.com/13340320/122829820-db9bba00-d2bd-11eb-818c-c8fa47f12108.png)

What I did here is a change on the **validateConnection** element, where we check if the connection is actually alive (using the events object). Aside of this, I modified my database timeouts to handle only 10ms for idle connections

```javascript
validateConnection(connection) {
    return connection._events != null;
}
```

Also, I modify the end listener to end the connection after the reading in order to trigger the validate method on the tarn pool.

```javascript
connection.on('end', async () => {
  connection.end();
  connection.__knex__disposed = 'Connection terminated unexpectedly'; //optional here
});
```
If I print the error that comes from the Postgres engine when the instance is shut down, we see the mythical error that everyone saw:
![image](https://user-images.githubusercontent.com/13340320/122830337-94fa8f80-d2be-11eb-8136-7627934ef05c.png)

At least with my replication queries seems to work after the first fail (error on orange), but as I said, I'm not sure if this solve the core issue of this thread for idle statuses (I don't have time to wait for an hour for open connections)

**Comment by NelsonFrancisco** [comment]
@Warkanlock thank you so much for the insight and provided solution. As far as I can help, I've made some tests.

- I was previously using postgres 12, but instead made the tests with version 11, so we are on the same page;
- Same knex configuration;
- I ran the tests in a similar environment as you: making requests within an interval. I had more intervals, that's the only difference;

Results: exactly the same as you did, but the error I get before your patch is not "Connection terminated unexpectedly" but instead "terminating connection due to administrator command", I have no idea why. After the patch, I get no error and the queries run just fine.

So, later I changed to postgres 12, and I'm sorry to say that your fix does not work. It throws the error anyway, whereas with postgres 11 it keeps doing queries.

But, anyways, make the PR and explain your doubts. I'm sure people will test and give feedback as I did.

**Comment by Dakuan** [comment]
very much still an issue, seems to be happening to all M1 macs, suggested workarounds not resolving

**Comment by david-wb** [comment]
Does knex have any built-in retry behavior for queries that fail with **Connection terminated unexpectedly**. It seems like that should exist, because network issues can happen at any time. Basic retries with exponential backoff and a `maxRetries` config setting would be great.

**Comment by elhigu** [comment]
> @david-wb Does knex have any built-in retry behavior for queries that fail with Connection terminated unexpectedly? It seems like that should exist, because network issues can happen at any time. Basic retries with exponential backoff and a maxRetries config setting would be great.

No, the root cause should be fixed. Retrying is pretty much impossible, because there is no way to know what was that connections state when it happened. For example if there was transaction going on etc. Also your app should be always prepared that any of your SQL queries may fail at any moment. 

@Warkanlock and others so were you really able to reproduce this problem, killing db connections while knex was doing other stuff? 

I have tried earlier following methods: killing queries with pg_terminate_backend, shut down DB  and restart, using toxiproxy between knex and DB and closing connections with it while queries running at certain moment or at certain amount of bytes sent through proxy (allows disconnections at very weird moments during the queries). 

Some tests code that I was using for stress testing drivers and their error handling / memory leaks / connection leaks is found here https://github.com/knex/knex/tree/master/scripts/stress-test

If you are able to create any test case replicating this behavior there it would be really great.

**Comment by shousper** [comment]
@elhigu I get quite a lot of errors with that `stress-test` script after just a minute or two.

```
$ node knex-stress-test.js| grep -A 3 -- "- STATS "
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
PSQL: Error: Connection terminated unexpectedly
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
PSQL TO: Error: connect ETIMEDOUT 127.0.0.1:25432
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
PSQL: Error: Connection terminated unexpectedly
(node:45754) UnhandledPromiseRejectionWarning: Error: Proxy postgresql already exists
    at Toxiproxy.<anonymous> (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:31:27)
    at Generator.throw (<anonymous>)
    at rejected (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:5:65)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:45754) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 40365)
(node:45754) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
(node:45754) UnhandledPromiseRejectionWarning: Error: Proxy postgresql already exists
    at Toxiproxy.<anonymous> (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:31:27)
    at Generator.throw (<anonymous>)
    at rejected (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:5:65)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
(node:45754) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 40402)
(node:45754) UnhandledPromiseRejectionWarning: Error: Response status was not 200: 409
    at Proxy.<anonymous> (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Proxy.js:109:23)
    at Generator.throw (<anonymous>)
    at rejected (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Proxy.js:5:65)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
(node:45754) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 40403)
(node:45754) UnhandledPromiseRejectionWarning: Error: Proxy postgresql already exists
    at Toxiproxy.<anonymous> (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:31:27)
    at Generator.throw (<anonymous>)
    at rejected (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:5:65)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
(node:45754) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 41391)
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 11.5, results: 8, errors: 3.5 } }
------------------------------- EOS ------------------------------
KILLER ERROR: Error: Connection terminated unexpectedly
(node:45754) UnhandledPromiseRejectionWarning: Error: Proxy postgresql already exists
    at Toxiproxy.<anonymous> (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:31:27)
    at Generator.throw (<anonymous>)
    at rejected (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:5:65)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
(node:45754) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 43664)
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
KILLER ERROR: Error: Connection terminated unexpectedly
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 8.5, results: 5.5, errors: 3 } }
------------------------------- EOS ------------------------------
KILLER ERROR: Error: Connection terminated unexpectedly
(node:45754) UnhandledPromiseRejectionWarning: Error: Proxy postgresql already exists
    at Toxiproxy.<anonymous> (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:31:27)
    at Generator.throw (<anonymous>)
    at rejected (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:5:65)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
(node:45754) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 48523)
(node:45754) UnhandledPromiseRejectionWarning: Error: Proxy postgresql already exists
    at Toxiproxy.<anonymous> (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:31:27)
    at Generator.throw (<anonymous>)
    at rejected (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:5:65)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
(node:45754) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 51548)
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
PSQL TO: Error: Connection terminated unexpectedly
(node:45754) UnhandledPromiseRejectionWarning: Error: Proxy postgresql already exists
    at Toxiproxy.<anonymous> (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:31:27)
    at Generator.throw (<anonymous>)
    at rejected (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:5:65)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
(node:45754) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 54765)
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
PSQL TO: Error: Connection terminated unexpectedly
(node:45754) UnhandledPromiseRejectionWarning: Error: Proxy postgresql already exists
    at Toxiproxy.<anonymous> (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:31:27)
    at Generator.throw (<anonymous>)
    at rejected (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:5:65)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
(node:45754) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 55706)
(node:45754) UnhandledPromiseRejectionWarning: Error: Proxy postgresql already exists
    at Toxiproxy.<anonymous> (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:31:27)
    at Generator.throw (<anonymous>)
    at rejected (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:5:65)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
(node:45754) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 55707)
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
KILLER ERROR: Error: Connection terminated unexpectedly
(node:45754) UnhandledPromiseRejectionWarning: Error: Proxy postgresql already exists
    at Toxiproxy.<anonymous> (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:31:27)
    at Generator.throw (<anonymous>)
    at rejected (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:5:65)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
(node:45754) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 57520)
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
KILLER ERROR: Error: Connection terminated unexpectedly
(node:45754) UnhandledPromiseRejectionWarning: Error: Proxy postgresql already exists
    at Toxiproxy.<anonymous> (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:31:27)
    at Generator.throw (<anonymous>)
    at rejected (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:5:65)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
(node:45754) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 64802)
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 5.5, results: 3, errors: 2.5 } }
------------------------------- EOS ------------------------------
PSQL TO: Error: Connection terminated unexpectedly
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
PSQL: Error: Connection terminated unexpectedly
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 11.5, results: 7, errors: 4.5 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
(node:45754) UnhandledPromiseRejectionWarning: Error: Proxy postgresql already exists
    at Toxiproxy.<anonymous> (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:31:27)
    at Generator.throw (<anonymous>)
    at rejected (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:5:65)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
(node:45754) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 79412)
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
PSQL: Error: Connection terminated unexpectedly
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
KILLER ERROR: Error: Connection terminated unexpectedly
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
PSQL: Error: Connection terminated unexpectedly
(node:45754) UnhandledPromiseRejectionWarning: Error: Proxy postgresql already exists
    at Toxiproxy.<anonymous> (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:31:27)
    at Generator.throw (<anonymous>)
    at rejected (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:5:65)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
(node:45754) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 86561)
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 3, results: 0, errors: 3 } }
------------------------------- EOS ------------------------------
KILLER ERROR: Error: Connection terminated unexpectedly
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 11, results: 6.5, errors: 4.5 } }
------------------------------- EOS ------------------------------
PSQL TO: Error: Connection terminated unexpectedly
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
PSQL: Error: Connection terminated unexpectedly
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 3.5, results: 2, errors: 1.5 } }
------------------------------- EOS ------------------------------
KILLER ERROR: Error: Connection terminated unexpectedly
(node:45754) UnhandledPromiseRejectionWarning: Error: Response status was not 200: 404
    at Proxy.<anonymous> (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Proxy.js:109:23)
    at Generator.throw (<anonymous>)
    at rejected (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Proxy.js:5:65)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
(node:45754) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 99436)
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 14.5, results: 10, errors: 4.5 } }
------------------------------- EOS ------------------------------
PSQL TO: Error: Connection terminated unexpectedly
(node:45754) UnhandledPromiseRejectionWarning: Error: Proxy postgresql already exists
    at Toxiproxy.<anonymous> (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:31:27)
    at Generator.throw (<anonymous>)
    at rejected (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:5:65)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
(node:45754) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 108585)
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 2.5, results: 1, errors: 1.5 } }
------------------------------- EOS ------------------------------
Connection Error: error: terminating connection due to administrator command
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 393.5, results: 339, errors: 54.5 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
KILLER ERROR: Error: Connection terminated unexpectedly
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
PSQL: Error: Connection terminated unexpectedly
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
PSQL TO: Error: Connection terminated unexpectedly
(node:45754) UnhandledPromiseRejectionWarning: Error: Response status was not 200: 404
    at Proxy.<anonymous> (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Proxy.js:109:23)
    at Generator.throw (<anonymous>)
    at rejected (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Proxy.js:5:65)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
(node:45754) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 122042)
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 2.5, results: 1, errors: 1.5 } }
------------------------------- EOS ------------------------------
PSQL TO: Error: Connection terminated unexpectedly
(node:45754) UnhandledPromiseRejectionWarning: Error: Proxy postgresql already exists
    at Toxiproxy.<anonymous> (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:31:27)
    at Generator.throw (<anonymous>)
    at rejected (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:5:65)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
(node:45754) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 126079)
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
KILLER ERROR: Error: Connection terminated unexpectedly
(node:45754) UnhandledPromiseRejectionWarning: Error: Proxy postgresql already exists
    at Toxiproxy.<anonymous> (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:31:27)
    at Generator.throw (<anonymous>)
    at rejected (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:5:65)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
(node:45754) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 130461)
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 9, results: 4.5, errors: 4.5 } }
------------------------------- EOS ------------------------------
PSQL: Error: Connection terminated unexpectedly
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
PSQL TO: Error: Connection terminated unexpectedly
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 417, results: 389, errors: 28 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
PSQL TO: Error: Connection terminated unexpectedly
(node:45754) UnhandledPromiseRejectionWarning: Error: Proxy postgresql already exists
    at Toxiproxy.<anonymous> (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:31:27)
    at Generator.throw (<anonymous>)
    at rejected (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:5:65)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
(node:45754) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 132374)
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
PSQL: Error: Connection terminated unexpectedly
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 3.5, results: 2, errors: 1.5 } }
------------------------------- EOS ------------------------------
PSQL TO: Error: Connection terminated unexpectedly
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 477.5, results: 429, errors: 48.5 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
KILLER ERROR: Error: Connection terminated unexpectedly
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
PSQL TO: Error: Connection terminated unexpectedly
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
PSQL TO: Error: Connection terminated unexpectedly
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
PSQL TO: Error: Connection terminated unexpectedly
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
PSQL: Error: Connection terminated unexpectedly
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 5, results: 2.5, errors: 2.5 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
PSQL: Error: Connection terminated unexpectedly
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 0, results: 0, errors: 0 } }
------------------------------- EOS ------------------------------
PSQL TO: Error: Connection terminated unexpectedly
(node:45754) UnhandledPromiseRejectionWarning: Error: Proxy postgresql already exists
    at Toxiproxy.<anonymous> (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:31:27)
    at Generator.throw (<anonymous>)
    at rejected (/Users/cmcgregor/Source/public/knex/node_modules/toxiproxy-node-client/dist/Toxiproxy.js:5:65)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
(node:45754) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 158028)
--
------------------------ STATS PER SECOND ------------------------
{ pg: { queries: 3.5, results: 1.5, errors: 2 } }
------------------------------- EOS ------------------------------
PSQL: Error: Connection terminated unexpectedly
^C
```

I cut down the script to only include `pg` tests and the docker compose file to only run a `postgres:12.4-alpine` image container to match what we're using. I also noticed `knex` doesn't include a `package-lock.json` file anywhere in git, so running against `master` isn't necessarily a true representation of how a release version will behave..

Can I ask, what kind of output indicates nominal or problematic behaviour? The errors here are very much what I've seen during a standby failover (via RDS, or RDS Proxy) though `knex` sometimes doesn't manage to recover. It'll get stuck in a weird state where every or most connections report these kinds of errors until the entire `knex` client is destroyed & reinitialised (or the service is restarted). Without sticking my fingers into internals of `knex` and `tarn`, touching properties I probably shouldn't, it's quite difficult to instrument adequately to chart behaviour via logs or metrics.

Unrelated note.. is there any plan to support self-management of connections & pooling? I think I've seen a few people say they've solve their issues using `pg` and/or `pg-pool` directly (sacrificing use of knex), and I'd imagine that to be a desirable escape hatch for some folks.

**Comment by priitkaard** [comment]
I haven't followed the progress of this issue, but for those who arrive here searching for solutions to the `Connection ended unexpectedly` errors - 

the reason is that `node-postgres` [disables keepAlive](https://github.com/brianc/node-postgres/blob/684cd09bcecbf5ad5f451fdf608a3e9a9444524e/packages/pg/lib/client.js#L51) by default. 
In our case, TCP connections idle for more than 15min got killed and the first query after 15min always failed with `Connection ended unexpectedly`

A workaround is to set `keepAlive` to true:
```typescript
const knex = Knex({
    client: 'pg',
    connection: {
        host: '',
        user: '',
        password: '',
        keepAlive: true,
    },
    pool: {
        min: 2,
        max: 20
    }
})
```

**Comment by timmeh87** [comment]
I am running knex inside a docker swarm and I have this issue so I am joining this thread. I am in development right now so I get this error every morning, I am the only person making requests.

I dont have time for this, so I made a workaround. This morning, no error. This is my workaround, I understand that we want a real solution that addresses the issue but for now, if changing the connection pool size does not work, just make a request every 5 minutes. No error this morning. 🤷

```
try{
// knex.js has a known issue inside of docker swarm
//https://github.com/knex/knex/issues/3523
//if there is inactivity, the next DB connection after inactivity will error out
//the best solution provided on the issue was to set the pool size to zero
//but I thought 'why not just make a request every 5 minutes to keep it alive'
	setInterval( () =>
	{
		knex('users')
			.select('*')
			.where({_id:1})
			.then((data) =>
			{
				console.log("knex keepalive hack");
			})
	}, 5*60*1000)
} catch { console.log("knex keepalive hack failure");}
```

**Comment by jpike88** [comment]
Is there anything we can learn from pg-pool?

https://github.com/brianc/node-postgres/blob/master/packages/pg-pool/index.js

**Comment by slvlirnoff** [comment]
Could it be linked to this: https://github.com/moby/moby/issues/37466 ?
workaround here: https://github.com/moby/moby/issues/37466#issuecomment-405537713

**Comment by sshettyyara** [comment]
Is the issue still open i am getting bellow error with knex 0.21.21 and pg 8.7.1 sometimes i production with the configuration {min: 2, max: 10}

Error: Connection terminated
    at Connection.<anonymous> (/app/node_modules/pg/lib/client.js:132:36)
    at Object.onceWrapper (events.js:519:28)
    at Connection.emit (events.js:412:35)
    at Connection.emit (domain.js:475:12)
    at Socket.<anonymous> (/app/node_modules/pg/lib/connection.js:57:12)
    at Socket.emit (events.js:400:28)
    at Socket.emit (domain.js:475:12)
    at TCP.<anonymous> (net.js:686:12)
    at TCP.callbackTrampoline (internal/async_hooks.js:130:17)

**Comment by timmeh87** [comment]
In my server I have this comment, part of the initialization

    pool: { min: 0, max: 2 }  // IMPORTANT https://knexjs.org/guide/#pool

I also dont seem to have the problem these days, I think maybe the key is
min:0 ?

**Comment by sshettyyara** [comment]
It seems like the previous issue pertained to running Knex inside a Docker Swarm, but our application is running in Kubernetes. Will setting { min: 0 } also resolve the issue for us in this Kubernetes environment ?? 
within kubernetes yes its running as docker containers

**Comment by timmeh87** [comment]
The document linked in my previous post suggests that zero is the safest
setting and the default of 2 is "for legacy" so I would go with zero and do
some testing

**Comment by Warkanlock** [comment]
Coming here after a while, setting this to zero it's the most failover practice you could use on knex.js to avoid this kind of issue.

So far, using it on production for at least 20 different servers, and none of them is throwing the error anymore (after setting it to `pool : { min : 0, max: N }`

Maybe the issue has been resolved on tarn.js + knex library, but still, good to have it mind if you are experimenting this kind of issues
